### PR TITLE
pedersen: added zock pedersen hashing to a lib

### PR DESCRIPTION
**Pull request by tacryt-socryp**

### CODE REVIEW

File: lib/zink/zink.hoon [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+    [0 0 cache]
+  ::  pederson hash noun
+  ++  hash-noun
+    |=  n=*
+    ^-  [phash (map * phash)]
+    (~(hash eval-door [~ cache]) n)
+  ::  evaluate nock with zink
+  ++  eval-noun
+    |=  n=^
+    ^-  [res=* hint-js=@t c=(map * phash)]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:22]: unnecessary to put labels on return types `res=` etc

File: lib/zink/zink.hoon [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+    =/  [res=* st=[h=hints c=(map * phash)]]
+          (~(eval eval-door [~ cache]) n)
+    =.  cache  c.st
+    =/  ch  (create-hints n h.st)
+    [res (crip (en-json:html js.ch)) cache.ch]
+  ::  compile a hoon file and evaluate it with zink
+  ++  eval-hoon
+    |=  [lib=path file=path arm=@tas sample=@t]
+    ^-  [res=* hint-js=@t c=(map * phash)]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:31]: same

File: lib/zink/zink.hoon [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+    |=  [arm=@tas sample=@t]
+    ^-  hoon
+    [%cncl [%wing ~[arm]] ~[(ream sample)]]
+  ::  create full hint json
+  ++  create-hints
+    |=  [n=^ h=hints]
+    ^-  [js=json cache=(map * phash)]
+    =^  hs  cache  (~(hash eval-door [~ cache]) -.n)
+    =^  hf  cache  (~(hash eval-door [~ cache]) +.n)
+    :-  %-  pairs:enjs:format
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:51]: use `:_` when the second argument is much shorter than the first

File: lib/zink/zink.hoon [hunk 4]
```diff
@@ -0,0 +1,353 @@
...
+        :~
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:52]:
```
:~  ['subject' ...]
```

File: lib/zink/zink.hoon [hunk 5]
```diff
@@ -0,0 +1,353 @@
...
+          ['formula' s+(num:enjs hf)]
+          ['hints' (all:enjs h)]
+        ==
+    cache
+  ::
+  ++  eval-door
+    |_  st=[h=hints c=(map * phash)]
+    ++  eval
+      |=  [s=* f=*]
+      ^-  [res=* st=[h=hints c=(map * phash)]]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:63]: same

File: lib/zink/zink.hoon [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+      =*  h  h.st
+      =^  sroot  c  (hash s)
+      =^  froot  c  (hash f)
+      |^
+      ::~&  >  "{<s>}  {<f>}"
+      ?+    -.f  !!
+        ::  formula is a cell; do distribution
+        ::
+          ^
+        =/  [subf1=* subf2=*]  [-.f +.f]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:74]: Use `=*` twice instead of adding to the subject

File: lib/zink/zink.hoon [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+        =^  res-head  st
+          (eval [s subf1])
+        =^  res-tail  st
+          (eval [s subf2])
+        =^  hsubf1  c  (hash subf1)
+        =^  hsubf2  c  (hash subf2)
+        :-  [res-head res-tail]
+            [(put-hint [%cons hsubf1 hsubf2]) c]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:82]: dedent 82 to match 81

File: lib/zink/zink.hoon [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+        ::
+          %0
+        ?>  ?=(@ +.f)
+        =/  res  .*(s f)
+        =/  axis  +.f
+        =^  hres  c  (hash res)
+        =^  sibs  c  (merk-sibs s axis)
+        :-  res
+            [(put-hint [%0 axis hres sibs]) c]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:91]: same as 82

File: lib/zink/zink.hoon [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+        ::
+          %1
+        =^  hres  c  (hash +.f)
+        [+.f (put-hint [%1 hres]) c]
+        ::
+          %2
+        =/  [subf1=* subf2=*]  [+<.f +>.f]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:98]: `=*`

File: lib/zink/zink.hoon [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+        =^  hsubf1  c  (hash subf1)
+        =^  hsubf2  c  (hash subf2)
+        =^  hsubf3  c  (hash subf3)
+        =^  res1  st
+          (eval [s subf1])
+        ?>  ?|(=(0 res1) =(1 res1))
+        =^  res2  st
+          ?:  =(0 res1)
+            (eval [s subf2])
+            (eval [s subf3])
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:147]: dedent by one

File: lib/zink/zink.hoon [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+        [res2 (put-hint [%6 hsubf1 hsubf2 hsubf3]) c]
+        ::
+          %7
+        =/  [subf1=* subf2=*]  [+<.f +>.f]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:151]: `=*`

File: lib/zink/zink.hoon [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+        =^  hsubf1  c  (hash subf1)
+        =^  hsubf2  c  (hash subf2)
+        =^  res1  st
+          (eval [s subf1])
+        =^  res2  st
+          (eval [res1 subf2])
+        [res2 (put-hint [%7 hsubf1 hsubf2]) c]
+        ::
+          %8
+        =/  [subf1=* subf2=*]  [+<.f +>.f]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:161]: `=*`

File: lib/zink/zink.hoon [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+        =/  [subf1=* subf2=*]  [-.f +.f]
+        =^  res-head  st
+          (eval [s subf1])
+        =^  res-tail  st
+          (eval [s subf2])
+        =^  hsubf1  c  (hash subf1)
+        =^  hsubf2  c  (hash subf2)
+        :-  [res-head res-tail]
+            [(put-hint [%cons hsubf1 hsubf2]) c]
+        ::
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:83]: all of the comment clauses between conditionals in a switch statement should be dedented one further

File: lib/zink/zink.hoon [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+        =^  hsubf1  c  (hash subf1)
+        =^  hsubf2  c  (hash subf2)
+        =^  res1  st
+          (eval [s subf1])
+        =^  res2  st
+          (eval [[res1 s] subf2])
+        [res2 (put-hint [%8 hsubf1 hsubf2]) c]
+        ::
+          %9
+        =/  [axis=* subf1=*]  [+<.f +>.f]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:171]: `=*`

File: lib/zink/zink.hoon [hunk 15]
```diff
@@ -0,0 +1,353 @@
...
+          (eval [s subf1])
+        =^  f2  st
+          (eval [res1 [0 axis]])
+        =^  res2  st
+          (eval [res1 f2])
+        =^  hf2  c  (hash f2)
+        =^  sibs  c  (merk-sibs res1 axis)
+        =^  hres1  c  (hash res1)
+        :-  res2
+            [(put-hint [%9 axis hsubf1 hf2 sibs]) c]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:184]: dedent until flush with 183

File: lib/zink/zink.hoon [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+        ::
+          %10
+        =/  [axis=* subf1=* subf2=*]  [+<-.f +<+.f +>.f]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:187]: `=*`

File: lib/zink/zink.hoon [hunk 17]
```diff
@@ -0,0 +1,353 @@
...
+          (eval [s subf1])
+        =^  oldtree  st
+          (eval [s subf2])
+        =/  res  .*(s f)
+        =^  oldleaf  st
+          (eval [oldtree 0 axis])
+        =^  holdleaf  c  (hash oldleaf)
+        =^  sibs  c  (merk-sibs oldtree axis)
+        :-  res
+            [(put-hint [%10 axis hsubf1 hsubf2 holdleaf sibs]) c]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:201]: dedent

File: lib/zink/zink.hoon [hunk 18]
```diff
@@ -0,0 +1,353 @@
...
+        ::
+          %11
+        =/  subf1=*  +>.f
+        (eval [s subf1])
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:205]: get rid of 204

File: lib/zink/zink.hoon [hunk 19]
```diff
@@ -0,0 +1,353 @@
...
+        =|  path=(list phash)
+        |-  ^-  [(list phash) (map * phash)]
+        ?:  =(1 axis)
+          [path c]
+        ?~  axis  !!
+        ?@  s  !!
+        =/  pick  (cap axis)
+        =^  sibling=phash  c
+          %-  hash
+             ?-(pick %2 +.s, %3 -.s)
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:229]: nasty, dedent until flush with 228

File: lib/zink/zink.hoon [hunk 20]
```diff
@@ -0,0 +1,353 @@
...
+      =/  mh  (~(get by c) n)
+      ?.  ?=(~ mh)
+        [u.mh c]
+      ?@  n
+        =/  h  (hash:pedersen n 0)
+        [h (~(put by c) n h)]
+      =^  hh  c  $(n -.n)
+      =^  ht  c  $(n +.n)
+      =/  h  (hash:pedersen hh ht)
+        [h (~(put by c) n h)]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:253]: No need to declare h, just go ahead and use `:-`

File: lib/zink/zink.hoon [hunk 21]
```diff
@@ -0,0 +1,353 @@
...
+    ::  where h = pedersen hash
+    ++  hash
+      |=  [n=*]
+      ^-  [phash (map * phash)]
+      =*  c  c.st
+      =/  mh  (~(get by c) n)
+      ?.  ?=(~ mh)
+        [u.mh c]
+      ?@  n
+        =/  h  (hash:pedersen n 0)
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:248]: same here

File: lib/zink/zink.hoon [hunk 22]
```diff
@@ -0,0 +1,353 @@
...
+  ++  all
+    |=  h=^hints
+    ^-  json
+    (hints h)
+  ++  hints
+    |=  h=^hints
+    |^  ^-  json
+    %-  pairs:enjs:format
+    %+  turn  ~(tap by h)
+      |=  [sroot=phash v=(map phash cairo-hint)]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:267]: dedent both 267 and 268 and then put a `::` after

File: lib/zink/zink.hoon [hunk 23]
```diff
@@ -0,0 +1,353 @@
...
+      [(num sroot) (inner v)]
+    ++  inner
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:269]: `::` should go between all arms in every core

File: lib/zink/zink.hoon [hunk 24]
```diff
@@ -0,0 +1,353 @@
...
+      |=  i=(map phash cairo-hint)
+      ^-  json
+      %-  pairs:enjs:format
+      %+  turn  ~(tap by i)
+        |=  [froot=phash hin=cairo-hint]
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:274]: dedent 274 and 275

File: lib/zink/zink.hoon [hunk 25]
```diff
@@ -0,0 +1,353 @@
...
+      :-  %a
+      ^-  (list json)
+      ?-  -.hin
+          %0
+        :~  s+'0'
+            s+(num axis.hin)
+            s+(num leaf.hin)
+            a+(turn path.hin |=(p=phash s+(num p)))
+        ==
+        ::
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:288]: dedent all comments one further

File: lib/zink/zink.hoon [hunk 26]
```diff
@@ -0,0 +1,353 @@
...
+          %2
+        ~[s+'2' s+(num subf1.hin) s+(num subf2.hin)]
+        ::
+          %3
+        ::  if atom, head and tail are 0
+        ::
+        :*  s+'3'  s+(num subf.hin)
+            ?-  -.subf-res.hin
+                %atom
+              ~[s+(num +.subf-res.hin) s+'0' s+'0']
```
Comment (tacryt-socryp) [inline lib/zink/zink.hoon:301]: throw a `::` in between

File: lib/zink/zink.hoon (hunk 6 above)
Comment (martyr-binbex) [reply lib/zink/zink.hoon:74]: What is the reasoning here? Never use =/ for a wing because it's inefficient compared to =* which is just shorthand?

Comment (martyr-binbex) [reply lib/zink/zink.hoon:74]: i like that =/ lets you decompose the noun into multiple parts on one line vs using =* which takes multiple lines

File: lib/zink/zink.hoon (hunk 20 above)
Comment (martyr-binbex) [reply lib/zink/zink.hoon:253]: I need h because I also put it into c (the cache) in the tail of the cell

File: lib/zink/zink.hoon (hunk 21 above)
Comment (martyr-binbex) [reply lib/zink/zink.hoon:248]: I need this for the same reason (return it and also put it in the cache)

**Pull request merged**